pmd/wrapper/src/main/java/io/buildfoundation/bazel/pmd/Main.java

```diff
@@ -6,48 +6,136 @@
 import net.sourceforge.pmd.renderers.TextRenderer;
 
 import java.io.BufferedReader;
+import java.io.BufferedWriter;
 import java.io.FileReader;
+import java.io.FileWriter;
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
 
 public final class Main {
 
+    /**
+     * Entry point of the application.
+     * Processes input arguments, runs PMD, prints errors, and writes the execution result to a file.
+     */
     public static void main(String[] args) {
-        PMD.StatusCode result = PMD.runPmd(args);
+        List<String> inputArgs = Arrays.asList(args);
+        String executionResultOutputPath = getExecutionResultOutputPath(inputArgs);
+
+        String[] pmdArgs = sanitizePmdArguments(inputArgs);
+
+        PMD.StatusCode result = PMD.runPmd(pmdArgs);
 
         if (!result.equals(PMD.StatusCode.OK)) {
-            printError(args);
+            printError(pmdArgs);
         }
 
-        System.exit(result.toInt());
-    }
+        writeExecutionResultToFile(result, executionResultOutputPath);
 
-    private static void printError(String[] args) {
-        List<String> arguments = Arrays.asList(args);
+        System.exit(0);
+    }
 
-        String reportFormat = argument(arguments, "--format");
-        String reportFilePath = argument(arguments, "--report-file");
+    /**
+     * Prints the error report to the console if the specified report format and file path are provided.
+     */
+    private static void printError(String[] pmdArgs) {
+        List<String> pmdArguments = Arrays.asList(pmdArgs);
+        String reportFormat = getArgument(pmdArguments, "--format");
+        String reportFilePath = getArgument(pmdArguments, "--report-file");
 
         if (reportFormat != null && reportFilePath != null) {
-            if (Arrays.asList(TextRenderer.NAME, TextColorRenderer.NAME, TextPadRenderer.NAME).contains(reportFormat)) {
+            List<String> supportedFormats = Arrays.asList(TextRenderer.NAME, TextColorRenderer.NAME, TextPadRenderer.NAME);
+            if (supportedFormats.contains(reportFormat)) {
                 printFile(reportFilePath);
             }
         }
     }
 
-    private static String argument(List<String> arguments, String name) {
+    /**
+     * Writes the execution result to a file
+     */
+    private static void writeExecutionResultToFile(PMD.StatusCode statusCode, String executionResultOutputPath) {
+        String content;
+
+        content = String.format("#!/bin/bash\n\nexit %d\n", statusCode.toInt());
+
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(executionResultOutputPath))) {
+            writer.write(content);
+        } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Exception should be re-thrown here I think, otherwise we're exiting function without error and just text in error log